python/src/iceberg/types.py

```diff
@@ -59,134 +62,176 @@ class IcebergType:
         'IcebergType()'
     """
 
-    @property
-    def string_type(self) -> str:
-        return self.__repr__()
-
-    def __str__(self) -> str:
-        return self.string_type
+    @classmethod
+    def __get_validators__(cls):
+        # one or more validators may be yielded which will be called in the
+        # order to validate the input, each validator will receive as an input
+        # the value returned from the previous validator
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, v):
```

Comment:
Why is the parsing done in the validate method?

Reply:
Good one, I've moved this into the Types itself.

Reply:
Looks like this is still here?

Reply:
@Fokko, is this supposed to be left in?

Reply:
I'm not sure why I missed this comment. So, this is supposed to be still in, but I moved the actual parsing (regex) to the type itself. Because the schema is a `List[IcebergType]` it will try to validate that one first. As an example fixed[22]` It will not be passed to the Fixed validate method (because the `__root__` is dynamic).

Before I had everything in the validate method, now I call a static method on the class itself:

```python
            if v.startswith("fixed"):
                return FixedType.parse(v)
```

Reply:
Okay, I see the calls to `parse`. That brings us back to my original question, which is why are we doing parsing in a validate method? Shouldn't this validate the input rather than returning the deserialized value?

I would expect this to do something like the schema validator, which would change `"fixed[12]"` into `{"__root__": "fixed", "length": 12} or something.

Reply:
Doesn't look like the `parse_obj`/`parse_raw` gets invoked. The `IcebergType` is a bit of a special case, since it cannot be invoked, I think it would be okay to (mis)use the validator there. It gets called because it is part of the schema:
```
class Schema:
    fields: Tuple[IcebergType. ...]
```
When deserializing, it will first visit the IcebergType for validation. We can then handle the special cases for the decimal and fixed.